nemo/collections/nlp/data/machine_translation/machine_translation_dataset.py

```diff
@@ -0,0 +1,213 @@
+        clean=False,
+        max_seq_length=512,
+        min_seq_length=1,
+        max_seq_length_diff=512,
+        max_seq_length_ratio=512,
+    ):
+
+        self.src_tokenizer = tokenizer_src
+        self.tgt_tokenizer = tokenizer_tgt
+        self.tokens_in_batch = tokens_in_batch
+
+        src_ids = dataset_to_ids(dataset_src, tokenizer_src)
+        tgt_ids = dataset_to_ids(dataset_tgt, tokenizer_tgt)
+        if clean:
+            src_ids, tgt_ids = self.clean_src_and_target(
+                src_ids,
+                tgt_ids,
+                max_tokens=max_seq_length,
+                min_tokens=min_seq_length,
+                max_tokens_diff=max_seq_length_diff,
+                max_tokens_ratio=max_seq_length_ratio,
+            )
+        self.batch_indices = self.pack_data_into_batches(src_ids, tgt_ids)
+        self.batches = self.pad_batches(src_ids, tgt_ids, self.batch_indices)
+
+    def __len__(self):
+        return len(self.batches)
+
+    def __getitem__(self, idx):
+        src_ids = self.batches[idx]["src"]
+        tgt = self.batches[idx]["tgt"]
+        labels = tgt[:, 1:]
+        tgt_ids = tgt[:, :-1]
+        src_mask = (src_ids != self.src_tokenizer.pad_id).astype(np.int32)
+        tgt_mask = (tgt_ids != self.tgt_tokenizer.pad_id).astype(np.int32)
+        sent_ids = np.array(self.batch_indices[idx])
+        return src_ids, src_mask, tgt_ids, tgt_mask, labels, sent_ids
+
+    def pad_batches(self, src_ids, tgt_ids, batch_indices):
+        """
+        Augments source and target ids in the batches with padding symbol
+        to make the lengths of all sentences in the batches equal.
+        """
+
+        batches = {}
+        for batch_idx, b in enumerate(batch_indices):
+            src_len = max([len(src_ids[i]) for i in b])
+            tgt_len = max([len(tgt_ids[i]) for i in b])
+            src_ids_ = self.src_tokenizer.pad_id * np.ones((len(b), src_len), dtype=np.int)
+            tgt_ids_ = self.tgt_tokenizer.pad_id * np.ones((len(b), tgt_len), dtype=np.int)
+            for i, sentence_idx in enumerate(b):
+                src_ids_[i][: len(src_ids[sentence_idx])] = src_ids[sentence_idx]
+                tgt_ids_[i][: len(tgt_ids[sentence_idx])] = tgt_ids[sentence_idx]
+            batches[batch_idx] = {"src": src_ids_, "tgt": tgt_ids_}
+        return batches
+
+    def pack_data_into_batches(self, src_ids, tgt_ids):
+        """
+        Takes two lists of source and target sentences, sorts them, and packs
+        into batches to minimize the use of padding tokens. Returns a list of
+        batches where each batch contains indices of sentences included into it
+        """
+
+        # create buckets sorted by the number of src tokens
+        # each bucket is also sorted by the number of tgt tokens
+        buckets = {}
+        for i, src_id in enumerate(src_ids):
+            src_len, tgt_len = len(src_id), len(tgt_ids[i])
+            if src_len not in buckets.keys():
```

Comment:
Should be better to do `src_len not in buckets` no?

Reply:
Thanks, @MaximumEntropy! Fixed it.